Update Opticks to support Optimizely's JS SDK v5.3.2

vlacerda opened the pull request:
https://app.shortcut.com/findhotel/story/119714/upgrade-optimizely-sdk-and-opticks-to-latest-version

Commit:
Update Opticks to support Optimizely's JS SDK v5.3.2

Commit:
Updated README

Gerrit88 commented on packages/lib/src/core/booleanToggle.test.ts:
I still see some references to `booleanToggle`:
- Documentation: `docs/booleanToggles.md` and `docs/dead-code-removal.md`
- Logic: `packages/lib/src/integrations/simple.ts`
- Type: `BooleanToggleType`

Can we clean this up as well? If not, let's mark them as deprecated and create a follow-up story.

vlacerda replied:
Sure thing. Good catch. I cleaned them up. I think I managed to do so .

Gerrit88 commented on README.md:
```diff
@@ -14,24 +14,18 @@ The library consists of two related concepts:
 
-- `toggle` toggles that switch between multiple experiment variants (a/b/c/...)
-- `booleanToggle` toggles that turn functionality on or off (feature flags)
+A toggle allow you to switch between multiple experiment variants (a/b/c/...)
```
```suggestion
A toggle allows you to switch between multiple experiment variants (a/b/c/...)
```

Gerrit88 commented on packages/lib/src/integrations/optimizely.test.ts:
```diff
@@ -5,14 +5,12 @@ import {
 // During the tests:
-// for booleanToggle 'foo' yields true and 'bar' yields false, unless forced
 // for toggle 'foo' yields 'b' and 'bar' yields 'a', unless forced
 import datafile from './__fixtures__/dataFile'
```
Do we need to update this `datafile` to match the new configuration format?

Gerrit88 commented on packages/lib/src/integrations/optimizely.test.ts:
```diff
@@ -26,7 +24,11 @@ import Optimizely, {
   // @ts-expect-error
   getEnabledFeaturesMock,
```
Looks like this mock is not used so we can remove the import (and maybe even the mock declaration).

Gerrit88 commented on packages/lib/src/integrations/optimizely.test.ts:
```diff
@@ -213,26 +200,23 @@ describe('Optimizely Integration', () => {
-      testAudienceSegmentationCacheBusting(booleanToggle, isFeatureEnabledMock)
 
-      it("Returns Optimizely's value when no arguments supplied using booleanToggle", () => {
+      it("Returns Optimizely's value when no arguments supplied using", () => {
```
```suggestion
      it("Returns Optimizely's value when no arguments supplied", () => {
```

Gerrit88 commented on packages/lib/src/integrations/optimizely.ts:
```diff
@@ -182,10 +182,10 @@ export const initialize = (
 ) =>
   optimizelyClient.notificationCenter.addNotificationListener(
-    NOTIFICATION_TYPES.DECISION,
+    OptimizelyLib.enums.NOTIFICATION_TYPES.DECISION,
```
With this change, we don't need the `NOTIFICATION_TYPES` declaration on line 19 anymore, right?

Gerrit88 commented on packages/lib/package.json:
Remember to bump the version of the `opticks` package here.

dale-french replied:
No need to do this manually - you can run `npx changeset` from the root dir which will generate the changeset. Once this PR is merged, another PR will be opened automatically with the version bump + changelog

[Example](https://github.com/viodotcom/opticks/pull/78/files)

dale-french replied:
@vlacerda let me know if you face any issues generating the changeset

Gerrit88 reviewed (changes requested):
Thanks, Vini!

I left a few suggestions.

I also did some QA and everything seems to work as expected.

luisfmsouza reviewed (commented):
I was trying to run this new version on the mobile app, but found something weird. Will comment here in case someoneelse also found and fixed it.

On the `package/mobile-app/package.json` file, I added the line below:

```json
"opticks": "github:viodotcom/opticks#sc-119714-upgrade-optimizely-sdk-and-opticks-to-latest",
```

After running the install command, this is how the opticks package folder looked like (without the code). Maybe this type of install is not supported?

<img width="213" alt="image" src="https://github.com/viodotcom/opticks/assets/4405147/c9bdeee3-c5e7-474f-b996-3af1cd219df5">

---

How do you suggest I install this branch version into the app to test it out?

Gerrit88 commented:
@luisfmsouza, I'm not sure the best way to run from GitHub. 

But I QA'ed by pointing to the repo on my local machine: 
```
"opticks": "../../../opticks/packages/lib",
```

vlacerda committed:
Update README.md

Co-authored-by: Gerrit Burger <Gerrit88@users.noreply.github.com>

vlacerda committed:
Apply suggestions from code review

Co-authored-by: Gerrit Burger <Gerrit88@users.noreply.github.com>

Commit:
Code review

Commit:
Code review

vlacerda commented:
> I was trying to run this new version on the mobile app, but found something weird. Will comment here in case someoneelse also found and fixed it.
> 
> On the `package/mobile-app/package.json` file, I added the line below:
> 
> ```json
> "opticks": "github:viodotcom/opticks#sc-119714-upgrade-optimizely-sdk-and-opticks-to-latest",
> ```
> 
> After running the install command, this is how the opticks package folder looked like (without the code). Maybe this type of install is not supported?
> 
> <img alt="image" width="213" src="https://private-user-images.githubusercontent.com/4405147/337216444-c9bdeee3-c5e7-474f-b996-3af1cd219df5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc2NzI5NzAsIm5iZiI6MTcxNzY3MjY3MCwicGF0aCI6Ii80NDA1MTQ3LzMzNzIxNjQ0NC1jOWJkZWVlMy1jNWU3LTQ3NGYtYjk5Ni0zYWYxY2QyMTlkZjUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDZUMTExNzUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmUzZjA1MWZmN2M4NzI5Yzk3MmQyZWRiM2NiZmI5MTQ2ZTU2NDEyMzJhODczNzMzZWNhMjYyYjZiN2QyYzcxNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Y0nXiRidcNE4Y4oNldG9fS7DFGaguSnc1Vq9KWniq2A">
> How do you suggest I install this branch version into the app to test it out?

@luisfmsouza I tested with Opticks locally. The steps are roughly:
- Clone opticks
- Go to `opticks/packages/lib` folder and run `yarn link`
- Go to the root of daedalus and run `yarn link {../relative-path-to-opticks/packages/lib}`
  - This will add the link you added yourself but with the local version 
- On the webpack.config of the package you want to test you then need to include opticks to be compiled as well by babel.
  In the property `include` [here](https://github.com/findhotel/daedalus/blob/a1a894a067f5e19ad4a9fe8dcd5082709c8c4ff3/packages/client/config/webpack.config.js#L537) add  `path.resolve(__dirname, '../../../../opticks/packages/lib')`

jopdeklein commented on packages/lib/src/integrations/__mocks__/@optimizely/optimizely-sdk.ts:
```diff
@@ -5,30 +5,23 @@ export const voidEventDispatcher = {
-    return [`${userId}-test-1`, `${userId}-test-2`]
-  }
+export const decideMock = jest.fn((toggleKey) => ({
+  enabled: toggleKey === "foo"
```
```suggestion
  enabled: toggleKey === 'foo'
```

Gerrit88 reviewed (commented):
Thanks for those changes, Vini!

It's probably best to do the fix I mentioned [here](https://github.com/FindHotel/daedalus/pull/9339#discussion_r1629209848) in this PR as well.

jopdeklein commented on README.md:
```diff
@@ -64,9 +58,9 @@ See the [Optimizely integration documentation](docs/optimizely-integration.md).
-as with `booleanToggle`, we specify multiple variants of which one is active at
-any time. By convention the variants are named `a` (control), `b`, `c` etc.
+Toggles can be used to implement a/b/c style testing and on/off values as well.
+We specify multiple variants of which one is active at any time.
```
```suggestion
We specify multiple variants of which only one is active at any time.
```

jopdeklein commented on README.md:
```diff
@@ -64,9 +58,9 @@ See the [Optimizely integration documentation](docs/optimizely-integration.md).
-Toggles can be used to implement a/b/c style testing, instead of on/off values
-as with `booleanToggle`, we specify multiple variants of which one is active at
-any time. By convention the variants are named `a` (control), `b`, `c` etc.
+Toggles can be used to implement a/b/c style testing and on/off values as well.
```
```suggestion
Toggles can be used to implement a/b/c style MVT testing and on/off feature flags as well.
```

jopdeklein commented on packages/lib/src/integrations/optimizely.ts:
```diff
@@ -244,33 +241,26 @@ export const isUserInRolloutAudience = (toggleId: ToggleIdType) => {
-  )
-
-  return isInAnyAudience || !isEveryoneElseRulePaused
+  return isInAnyAudience
```
Since we're not explicitly checking the last fallback rule as "everyone else", does this construct rely on the fact we always have our own explicit "everyone else" audience rule?

vlacerda replied:
It's supposed to stand on its own. If we don't have any audience rule set this means `isInAnyAudience` will be `false`. Which matches our expectations. If you don't have any rule set your experiment is always turned off.

Commit:
Remove boolean toggle remnants

Commit:
Code review

Gerrit88 reviewed (approved):
Thanks, Vini! 🚀
